### PR TITLE
[Layout] Rebuild `OptionList` with layout primitives

### LINKED ISSUE #7952: [Layout foundations] Rebuild `OptionList` with layout primitives

### PR DESCRIPTION
**Pull request by aveline**
### WHY are these changes introduced?

Fixes #7952 

### WHAT is this pull request doing?

Updates `OptionList` to use new layout primitives.

One small visual change with this update due to replacing `margin` with `padding` is a slight increase (`4px`) in spacing when there are multiple sections in an `OptionList`. I don't think the added complexity to keep this the same as current doesn't seem worth it, but happy to hear other thoughts.

**Update:** Added a `Bleed` to resolve the spacing change. Feels hacky but prevents the visual change.

### CODE REVIEW

**Review by laurkim — APPROVED** [review]
Tophatted all the stories for OptionList at the breakpoints locally and all except the one with sections look visually the same, but I know you mentioned the extra spacing in the description 👍  

It's probably worth confirming with @sarahill or @nayeobkim that the extra 4px is okay before merging this to the batch 3 branch.

File: polaris-react/src/components/OptionList/OptionList.tsx [hunk 1]
```diff
@@ -130,24 +129,28 @@ export function OptionList({
...
+            paddingBlockStart={isFirstOption ? undefined : '2'}
+          >
             {titleMarkup}
-            <ul
-              className={styles.Options}
-              id={`${id}-${sectionIndex}`}
-              role={role}
-            >
+            {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
+            {/* @ts-ignore OptionList allows string as role but Box doesn't. @TODO: remove string from OptionList in v11 */}
```
Comment (kyledurand) [inline polaris-react/src/components/OptionList/OptionList.tsx]: 👍  Let's just make sure this is added to the v11 road map / issue / PR, whatever we have going for that

Another thing I think you could do is cast `role as BoxProps['role']` or maybe `as unknown as BoxProps['role']` and that way you wouldn't need 4 lines of comments. I'm fine with leaving it as is, maybe uglier is better for visibility

File: polaris-react/src/components/OptionList/OptionList.tsx [hunk 2]
```diff
@@ -130,24 +129,28 @@ export function OptionList({
...
+          </Box>
         );
       })
     : null;
 
   return (
-    <ul className={styles.OptionList} role={role}>
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore OptionList allows string as role but Box doesn't. @TODO: remove string from OptionList in v11
+    <Box as="ul" role={role} padding="2">
```
Comment (kyledurand) [inline polaris-react/src/components/OptionList/OptionList.tsx]: It seems like this is where the added space is coming from. Can we use `isFirstOption` here to change the value like you did above?

**Review by kyledurand — COMMENTED** [review]
Code looks good but I think I'd prefer if we found a way to keep the same spacing. Plus 1 on waiting for UX before 🚢 ing

File: polaris-react/src/components/OptionList/OptionList.tsx [hunk 3]
```diff
@@ -130,24 +130,30 @@ export function OptionList({
...
-              role={role}
-            >
-              {optionsMarkup}
-            </ul>
-          </li>
+            <Bleed marginBlockStart={title ? undefined : '05'} marginInline="0">
+              <Box
+                as="ul"
+                id={`${id}-${sectionIndex}`}
+                role={role as BoxProps['role']}
```
Comment (laurkim) [inline polaris-react/src/components/OptionList/OptionList.tsx:143]: Nice 💯

File: polaris-react/src/components/OptionList/OptionList.tsx (hunk 2 above)
Comment (aveline) [reply polaris-react/src/components/OptionList/OptionList.tsx]: The added spacing only happens on sections where there's no title, that's the combination where there were overlapping margins.

I added a `Bleed` to handle that and resolve the change in spacing. It feels hacky but at least it prevents the visual regression.

File: polaris-react/src/components/OptionList/OptionList.tsx (hunk 3 above)
Comment (aveline) [reply polaris-react/src/components/OptionList/OptionList.tsx:143]: I think the uglier version with all the comments is better for visibility. But we can ship either one.

File: polaris-react/src/components/OptionList/OptionList.tsx (hunk 1 above)
Comment (aveline) [reply polaris-react/src/components/OptionList/OptionList.tsx]: There's a branch for v11 https://github.com/Shopify/polaris/pull/7597 so I'll go ahead and open a pr against it.

File: polaris-react/src/components/OptionList/OptionList.tsx (hunk 3 above)
Comment (laurkim) [reply polaris-react/src/components/OptionList/OptionList.tsx:143]: I do agree the one with all the comments was better for visibility to know to remove it but I'm okay with either version!

**Review by kyledurand — APPROVED** [review]
🎉

**Pull request merged**